bpo-44392: Add Py_GenericAlias to C API docs (#26724)
python/cpython: Doc/c-api/typehints.rst

.. highlight:: c

.. _typehintobjects:

Objects for Type Hinting
------------------------

Various built-in types for type hinting are provided.  Currently,
two types exist -- :ref:`GenericAlias <types-genericalias>` and
:ref:`Union <types-union>`.  Only ``GenericAlias`` objects are currently exposed
when including :file:`Python.h`.

.. c:function:: PyObject* Py_GenericAlias(PyObject *origin, PyObject *args)

   Return a :ref:`GenericAlias <types-genericalias>` object on success,
   ``NULL`` on failure.  Equivalent to calling the Python class
   :class:`types.GenericAlias`.  *origin* and *args* arguments set the
   ``GenericAlias``\ 's ``__origin__`` and ``__args__`` attributes respectively.

>    *origin* should be a type, and *args* **must** be a :c:type:`PyTupleObject`.

Reviewer: I don't think that's true; the `setup_ga` method automatically packs anything that's not a tuple into a 1-tuple. Python itself passes a non-tuple here: https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Objects/abstract.c#L182

Author: Thanks for the catch and pointing this out. I should read the code more carefully next time.

.. highlight:: c

.. _typehintobjects:

Objects for Type Hinting
------------------------

Various built-in types for type hinting are provided.  Currently,
two types exist -- :ref:`GenericAlias <types-genericalias>` and

> :ref:`Union <types-union>`.  Only ``GenericAlias`` objects are currently exposed
> when including :file:`Python.h`.

Reviewer 2: ```suggestion
:ref:`Union <types-union>`.  Only ``GenericAlias`` is exposed to C.
```

.. c:function:: PyObject* Py_GenericAlias(PyObject *origin, PyObject *args)

>    Return a :ref:`GenericAlias <types-genericalias>` object on success,
>    ``NULL`` on failure.  Equivalent to calling the Python class

Reviewer 2: ```suggestion
   Create a :ref:`GenericAlias <types-genericalias>` object.
   Equivalent to calling the Python class
```

   :class:`types.GenericAlias`.  The *origin* and *args* arguments set the
   ``GenericAlias``\ 's ``__origin__`` and ``__args__`` attributes respectively.
   *origin* should be a :c:type:`PyTypeObject*`, and *args* can be a
   :c:type:`PyTupleObject*` or any ``PyObject*``.  If *args* passed is
   not a tuple, a 1-tuple is automatically constructed and ``__args__`` is set
   to ``(args,)``.
   Minimal checking is done for the arguments, so the function will succeed even
   if *origin* is not a type.
   The ``GenericAlias``\ 's ``__parameters__`` attribute is constructed lazily

>    from ``__args__``.

Reviewer 2: ```suggestion
   from ``__args__``.
   On failure, an exception is raised and ``NULL`` is returned.
```

>    Here's an example of how most types in the standard library use
>    ``Py_GenericAlias``.  This allows a type to be subscripted::

Reviewer 2: ```suggestion
   Here's an example of how to make an extension type generic::
```

      static PyMethodDef my_obj_methods[] = {
          {"__class_getitem__", (PyCFunction)Py_GenericAlias, METH_O|METH_CLASS, "See PEP 585"}

>       }
> 

Reviewer 2: ```suggestion
          ...
      }
      static PyTypeObject MyObject_Type = {
          // The usual
          ...
          tp_methods = my_obj_methods,
          ...
      };
```
